### DISCUSSION TITLE
Is job creation done in the same transaction as everything else, and can be rolled back?

### DISCUSSION THREAD

**Discussion by jjb** [bensheldon/good_job#908] [Q&A] [answered]
(This might be a very basic question and I'm missing something in the docs - I thought this was part of the main premise of goodjob, but couldn't find the answer in the readme.)

does job enqueueing happen in the same transaction as everything else, and will be rolled back?

example

```ruby
# my_model.rb
transaction do
  admins.each do |admin|
    UserMailer.with(user: admin).notify_about_thing.deliver_later
  end
  user1.update!(foo: 'bar')
end
```

if `user1.update!` raises an exception, will all the emails be removed from the queue?

**Comment by bensheldon** [comment, marked as answer]
Yes, though there might be edge cases; for example, I'm not sure about the intricacies of `require_new` and `joinable` and Active Record's built-in transactions.

The simplest case should work:

```ruby
expect do
  ApplicationRecord.transaction do
    TestJob.perform_later
    raise ActiveRecord::Rollback
  end
end.not_to change(GoodJob::Job, :count)
```

**Reply by bensheldon** [reply]
@jjb also not a basic question. Before you asked I assumed but never asserted it 😓

**Reply by jjb** [reply]
great!
